File: gui/lib/shapes/knob.rb
require_relative 'shape'

class Knob < RoundShape
  DIAMETERS = { huge: 19.0, large: 12.7, medium: 10.0, small: 8.4, tiny: 7.0, }

  attr_reader :slug

  def initialize(size:, knob_color:, pointer_color:)
    super(DIAMETERS[size])
    @knob_color = knob_color
    @pointer_color = pointer_color
    @slug = Pathname("knob-#{size}")
  end

  def draw(canvas)
    pointer_width = radius / 8.0
    pointer_length = radius - pointer_width
    canvas.g(stroke: @pointer_color, fill: @knob_color) do |g|
      g.circle(r: radius, stroke: 'none')
      g.line(y2: -pointer_length, 'stroke-width' => pointer_width, 'stroke-linecap' => 'round')
    end
  end
end

File: panelgen/internal/control/knob.go
package control

import (
	"dhemery.com/DHE-Modules/panelgen/internal/svg"
)

func TinyKnob(knobColor, pointerColor svg.Color) Control {
	return knob("knob-tiny", TinyKnobDiameter, knobColor, pointerColor)
}

func SmallKnob(knobColor, pointerColor svg.Color) Control {
	return knob("knob-small", SmallKnobDiameter, knobColor, pointerColor)
}

func MediumKnob(knobColor, pointerColor svg.Color) Control {
	return knob("knob-medium", MediumKnobDiameter, knobColor, pointerColor)
}

func LargeKnob(knobColor, pointerColor svg.Color) Control {
	return knob("knob-large", LargeKnobDiameter, knobColor, pointerColor)
}

const (
	HugeKnobDiameter   = 19
	LargeKnobDiameter  = 12.7
	MediumKnobDiameter = 10
	SmallKnobDiameter  = 8.4
	SmallKnobRadius    = SmallKnobDiameter / 2.0
	TinyKnobDiameter   = 7
	TinyKnobRadius     = TinyKnobDiameter / 2.0
)

func knob(slug string, diameter float64, knobColor, pointerColor svg.Color) Control {
	radius := diameter / 2
	knob := svg.Circle{
		R:    radius,
		Fill: knobColor,
	}

	pointerWidth := radius / 8
	pointerLength := radius - pointerWidth
	pointer := svg.Line{
		Stroke:      pointerColor,
		StrokeWidth: pointerWidth,
		Y2:          -pointerLength,
		Cap:         "round",
	}
	frame := svg.GroupOf(knob, pointer)
	return Control{
		Frames:       map[string]svg.Element{slug: frame},
		DefaultFrame: frame,
	}
}
